Activation Fit Function for Muons

sf1919 opened the pull request:
**Description of work.**
Added a new Activation Fit function for Muon modelling and associated documentation.

**To test:**
1. Enable the model analysis tab and plot (see #32572 for details on how to do this)
2. Open Muon Analysis GUI
2. Change Instrument to EMU
3. Load runs 20901-18
4. Go to the Grouping Tab 
5. Click Guess Alpha and select run 20918
6. Go to fitting tab and add a DynamicKuboToyabe Fitting Function
7. Set the Field to 0 and Fix it (right click on the attribute and a menu should appear with Fix as one of the options)
8.  Set Delta to 0.35 and Nu to 0.01. Do not fix these.
8. Go to the Sequential Fitting Tab and select runs 20901-17 and then sequentially fit selected
9. Go to the Results Tab
10. Select Temp_Sample from the log results and then click Output Results
11. Open the Model fitting tab and change x axis to Temp_sample and y axis to Nu
12. Add an Activation Fit Function and fit (leave Unit as K). AttemptRate = 33.4 +- 3.6 us^-1 and Barrier = 589.7+-13.9 and ChiSquared should be about 3.75 or better.

Fixes #32579. 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

sf1919 committed:
Set up activation fit function header and souce files

Created initial header and source files for activation fir function and updated Cmake lists accordingly.

re #32579

sf1919 committed:
Revised source and header with revised function

Additional information provided about function which simplified the maths. Header and source files updated accordingly.

re #32579

sf1919 committed:
Added tests and updated source file

Added tests for Activation Function
Updated source file to handle units being entered in both upper and lower case

re #32579

sf1919 committed:
Updated parameters and added documentation

Updated parameter names and defaults
Added documentation for Activation fitting function and release note

ref #32579

sf1919 committed:
Fixed Jenkins compiler failures

re #32579

AnthonyLim23 commented on Framework/CurveFitting/src/Functions/Activation.cpp:
```diff
@@ -0,0 +1,84 @@
+void Activation::init() {
+  declareAttribute("Unit", Attribute("K")); // or meV
+  declareParameter("AttemptRate", 1000.0, "coefficient for attempt rate");
+  declareParameter("Barrier", 1000.0, "coefficient for barrier energy");
```
what unit? I assume meV

sf1919 replied:
The unit is either K or meV. Depending on which is chosen will affect how the fitting function works.

AnthonyLim23 replied:
Could you please write that?

AnthonyLim23 commented on Framework/CurveFitting/src/Functions/Activation.cpp:
```diff
@@ -0,0 +1,84 @@
+
+double Activation::getMeVConv() const {
+  auto unit = getAttribute("Unit").asString();
+  boost::to_lower(unit);
```
do we want to make a distinction between mev and Mev? Since Mev is an abbreviation for Mega electron volts. So this might allow a user to think they are using MeV when they are not.

AnthonyLim23 replied:
are we going to handle the MeV and meV cases?

sf1919 replied:
I have updated it now so it will only accept K and meV as the units. I've also updated the tests and documentation.

AnthonyLim23 commented on docs/source/fitting/fitfunctions/Activation.rst:
```diff
@@ -0,0 +1,48 @@
+
+An activation fitting function, when using Kelvin, may be used to describe:
+
+.. math:: y = A_R\times e^{-\frac{b}{x}}
```
dont need the times

AnthonyLim23 commented on docs/source/fitting/fitfunctions/Activation.rst:
```diff
@@ -0,0 +1,48 @@
+Description
+-----------
+
+An activation fitting function, when using Kelvin, may be used to describe:
```
```suggestion
The activation fitting function can be written as:
```

AnthonyLim23 commented on docs/source/fitting/fitfunctions/Activation.rst:
```diff
@@ -0,0 +1,48 @@
+where:
+- A_R - Attempt Rate
+- b - Barrier energy
+
```
```suggestion
- x is in Kelvin
```

AnthonyLim23 commented on docs/source/fitting/fitfunctions/Activation.rst:
```diff
@@ -0,0 +1,48 @@
+
+References
+----------
+[1] Gil, J.M et al (1999). Novel Muonium State in CdS. Phys. Rev. Lett., Vol 83 Issue 25, 5294-5297 `doi: 10.1103/PhysRevLett.83.5294 <https://doi.org/10.1103/PhysRevLett.83.5294>`.
```
this reference is for the smooth transition function see figure 4

AnthonyLim23 replied:
https://iopscience.iop.org/article/10.1088/2515-7639/ac22ba/meta equation 1 has A_R=1 and has taken the logs

AnthonyLim23 commented on docs/source/fitting/fitfunctions/Activation.rst:
```diff
@@ -0,0 +1,48 @@
+.. math:: y = A_R\times e^{-\frac{E\times b}{1000 k_B x}}
+
+where:
+E = energy spin
```
E is the activation energy

sf1919 committed:
Changes to header, source and documentation on review

Updated getmeVConv function name in header and source files
Added a description for Unit parameter
Updated maths and source in documentation

re #32579

sf1919 committed:
Force users to use K or meV for Units

Updated source, tests and documentation so that only K and meV will be accepted as Units.

re #32579

sf1919 committed:
Updated source file and documentation after review

Removed times from maths in documentation
Tidied source file to have correct units

re #32579

AnthonyLim23 commented:
I can crash mantid. If I do the fit with one attribute (say k), then change it to something wrong (e.g. K) and try to do a fit. I get an error message, thats ok. If I then change it aback to something correct and try and do a fit Matnid dies

sf1919 committed:
Remove need for boost in source file

re #32579

sf1919 commented:
> 
> 
> I can crash mantid. If I do the fit with one attribute (say k), then change it to something wrong (e.g. K) and try to do a fit. I get an error message, thats ok. If I then change it aback to something correct and try and do a fit Matnid dies

I get a crash because it doesn't have the correct Start and End X. I think thats because it was fixed in a PR that went in after this branch was started - #32857. I'm going to try rebasing it to see if that fixes it.

sf1919 committed:
Change MantidTestHelpers to MantidFrameworkTestHelpers in test

re #32579

AnthonyLim23 commented:
this still crashes for me. @Harrietbrown does it crash for you or have I got a bad build?

Harrietbrown commented:
I am getting the issue with the start_x and end_x 
![image](https://user-images.githubusercontent.com/49688535/141492133-27bc3149-de27-4f2c-b6bf-abefbcda1c2a.png)
if i mannualy change the start x and end x I can then get a fit to run

AnthonyLim23 commented:
I will have a look at that crash.

AnthonyLim23 reviewed (approved):
Looks good and adds a nice fit function
